active/0000-bicep-version-pinning.md

```diff
@@ -0,0 +1,231 @@
+
+*Disadvantages*:
+
+1. Could be considered more verbose in certain scenarios, e.g. to specify Bicep to match any patch version of `0.31`, the syntax would be `>=0.31.0 <0.32.0`. Likewise to match any minor version of `0`, the syntax would be `>=0 <1`.
+
+
+**Recommendation** - Due to the vast usage of the "npm/terraform syntax", and how many more tools support parsing this syntax, it seems like the better choice here. 
+
+
+**3. Wildcard syntax**
+
+Users can use `*` to indicate that anything in the specified segment of the version is acceptable. This can be used as syntactic sugar for the version comparison syntax. The following are some examples:
+- `1.*` - equivalent to `>=1.0.0,<2.0.0`
+- `1.2.*` - equivalent to `>=1.2.0,<1.3.0`
+- `*` -  equivalent to `>=0.0.0`
+
+However, we should impose the following restrictions:
+1. The wilcard `*` cannot be used together with the version comparison syntax. 
+    - Reason: This is to prevent ambiguous snytaxes like this: `>=1.*`. It is hard to reason about the allowed range in this case, i.e. does the range start from `1.0.0` or `1.1.2` or `1.3.5` or something else?
+2. If a version number is replaced with a wildcard, then all later version numbers must also be wildcards. Thus `2.*.6` is an invalid wildcard version. 
+    - Reason: Most semver parsing tools (including the [tools sampled above](#option-2-npmterraform-syntax)) either do not support this, or they ignore anything after the `*`.
+
+### VSCode experience
+Because the VSCode extension version is tightly coupled to the Bicep compiler version. If we fail the Bicep compilation, the user if forced to update their VSCode extension version; this is not a great user experience especially when working across multiple Bicep repos that they even might not own. 
+
+Ideally, we would decouple the extension from the Bicep version, but that would be an expensive undertaking.
+
+An acceptable solution here is: instead of throwing compiler errors, we'd warn the user if their VSCode extension version doesn't match the configured version constraint. 
+This way the user is aware of the mismatch, but we do not completely block them from proceeding with compilation.
+We will need to clearly document this behavior in which the pinning is not strictly enforced in VSCode, so that users do not have the wrong expectations.
+
+
+### AzCLI updates
+- The AzCLI bicep module should be enhanced to interpret the `bicepconfig.json` version constraint.
+- When a command that requires compilation is invoked (e.g. `build`, `build-params`, `lint`, `publish`, `restore`), find the appropriate `bicepconfig.json` file by recursively searching upwards from the current directory. 
+    - If a constraint is found, download the latest version that satisfies the constraint by introspecting Bicep release tags. For example, given the version constraint `>=0.31.0 <0.32.0` and the available Bicep release tags `[0.21.14, 0.31.1, 0.31.92, 0.32.45]`, AzCLI would download version `0.31.92`.
```

Comment:
Worth caching the versions e.g in `.azure/bin/<version>/bicep.exe` and copying to `.azure/bin/bicep.exe` when needed?